### ISSUE TITLE
Proposal: Endo[S] + Lens.then for write-side state invariants

### ISSUE THREAD

**Issue by n-gao** [cusp-ai-oss/kUPS#172]
## Summary

Add a write-side combinator to the lens library so a lens can re-establish a whole-state invariant after every write, without affecting reads:

- a type alias `Endo[S] = View[S, S]` (an endomorphism on the state), and
- a `then` combinator on `Lens`: `set(s, r) = after(inner.set(s, r))`, with `get` unchanged.

## Motivation

Variable-cell relaxation optimizes atomic positions and the cell together. When the optimizer writes a new cell, the atoms should ride with it — i.e. fractional coordinates are held fixed and Cartesian positions are recomputed as `positions = cell.to_real(fractional)`. Otherwise a box deformation throws every atom off its lattice site, which couples the position and cell subproblems badly.

The relaxation `property` is a plain lens onto `(positions, cell)`, so setting the cell leaves the Cartesian positions untouched and the invariant `positions == cell.to_real(fractional)` is silently violated. We want a lens whose *write* restores that invariant while its *read* stays a plain projection.

More generally: state updates often have derived/coupled fields that must be refreshed after a focused write. A write-side hook expresses this directly instead of scattering fix-ups across call sites or overloading the getter.

## Proposed API

```python
type Endo[S] = View[S, S]
"""An endomorphism on the state, post-composed with a lens's writes."""
```

On the `Lens` protocol (default-implemented in `BaseLens`, alongside `focus` / `merge` / `nest` / `at`):

```python
def then[S2](self: Lens[S2, R], after: Endo[S2]) -> Lens[S2, R]:
    """Re-establish a whole-state invariant after every write.

    Returns a lens with the same getter, whose setter post-composes the
    write with ``after``: ``set(s, r) == after(inner.set(s, r))``. The
    normalization is one-directional — reads are unaffected.
    """
```

`apply`, `at`, `merge`, and `nest`-derived writes inherit the behavior for free, since `BaseLens` routes them through `set`.

## Semantics and contract

`after` is an `Endo[S]` applied to the **post-set** state only — it never sees the pre-set state. So it must:

- depend only on the post-set state (store the independent DOF, e.g. fractional coords + cell, and derive the dependent one, e.g. Cartesian positions);
- be idempotent;
- write only fields disjoint from this lens's focus.

Under those conditions the lens laws hold on the consistent submanifold:

- **PutGet** holds because `after` touches only out-of-focus fields.
- **GetPut** holds whenever the state is already consistent (`after` is an idempotent fixpoint there).

Endomorphisms form a monoid under composition, so multiple invariants combine by composing their `Endo[S]` into one before passing to `then`.

## Example

```python
def add_lattice_dependence(s: State) -> State:
    cell = s.systems[s.particles.data.system].cell
    frac_pos = cell.to_fractional(s.particles.data.positions)
    real_pos = cell.to_real(jax.lax.stop_gradient(frac_pos))
    new_pos = x.particles.data.positions + (real_pos - jax.lax.stop_gradient(real_pos))
    return bind(s, lambda x: x.particles.data.positions).set(new_pos)

property = prop_lens.then(add_lattice_dependence)
```

## Scope / non-goals

`then` fixes the **primal** write: the state update and any energy re-evaluation through `property.set(...)` become consistent. The **gradient** handed to the optimizer arrives in Cartesian/partial coordinates `(∂E/∂r, ∂E/∂h|_r)` and must independently be expressed in the same `(fractional, cell)` coordinates (the matching cotangent: `∂E/∂s = Gʳ @ hᵀ`, `dE/dh|_s = ∂E/∂h|_r + h⁻ᵀ·Σ r⊗Gʳ`). That cotangent wiring is tracked separately; this issue covers only the lens combinator.

## Alternatives considered

- A pre/post hook `(S_old, S_new) -> S_new`: strictly more general, but heavier, and unnecessary once the independent DOF is stored and the dependent one derived.
- Domain-specific names (`normalize`, `reconcile`, `rectify`): rejected in favor of the generic `Endo` / `then`, which name the type and the composition rather than editorializing about intent.

**Comment by jonkhler** [comment]
@n-gao can you elaborate on the `positions_follow_cell` naming? like what is following what?

**Comment by n-gao** [comment]
I updated the example to be more reasonable/real-world. here we could attach this Endo to the gradient_lens to include the explicit dependence of the positions to the lattice vectors.

**Comment by jonkhler** [comment]
@n-gao the physics and the cotangent algebra are right. But this is a standard optic: a reversible coordinate change (`Iso`). Framing it that way collapses the primal and the gradient into one definition, with no submanifold caveat.

## It's an `Iso`
State stores `(positions, cell)`; the optimizer moves in `(fractional, cell)`; `r = cell.frame.to_real(fractional)` is invertible.

```python
@dataclass
class Iso[A, B]:
    fwd: Callable[[A], B]   # stored -> coords
    bwd: Callable[[B], A]   # coords -> stored

fractional_iso = Iso(
    fwd=lambda s: (s[1].frame.to_fractional(s[0]), s[1]),
    bwd=lambda c: (c[1].frame.to_real(c[0]),       c[1]),
)
```

## Both halves come from `bwd`
- primal ride-along: `bwd`
- coupled gradient: `jax.vjp(bwd)` on the Cartesian gradient. This *is* the two non-goals formulas, derived rather than hand-written:

```python
grad_coords = jax.vjp(bwd, coords)[1]((dE_dr, dE_dh))[0]
#  == (Gʳ @ hᵀ,  ∂E/∂h|_r + h⁻ᵀ Σ r⊗Gʳ)
```

No `stop_gradient`, no `from_matrix` (the VJP lands in `tril`-space). The potential is never differentiated; its analytic gradient is only the cotangent seed, so this holds for the torch-bridged ML potentials too.

## Lawful unconditionally
`bwd(fwd(x)) == x` and `fwd(bwd(x)) == x` for any non-degenerate cell. No idempotence requirement, no disjointness, no "consistent submanifold". That caveat is the cost of mutating a lens setter instead of using an iso.

## `Endo`/`then` vs `Iso`
- A mature optics library (python-lenses) has `Iso` (reversible) and `Prism` (partial) and **no** normalise-on-set optic. Its headline `Iso` is `lens.Json()` (serialize/parse); real/fractional is the same family.
- `Endo` names the wrong object: the contract requires idempotence, so it is an idempotent *retraction*, and retractions aren't closed under composition, so "Endos form a monoid, compose freely" doesn't hold.
- `positions_follow_cell` is just `bwd` (the direction of the iso), which answers the naming question from the thread above.

## Wiring
`RelaxationPropagator` keeps `property: Lens[State, Stored]` (*where*) and gains `reparam: Iso[Stored, Coords]` (*which coordinates*), default identity:

```python
coords = reparam.fwd(property.get(state))
grad   = jax.vjp(reparam.bwd, coords)[1](potential_out.data.gradients)[0]
...
state  = property.set(state, reparam.bwd(apply_updates(coords, updates)))
```

Identity iso gives byte-for-byte current behavior. `property` and `reparam` stay separate because the gradient is a `stored`/`coords` pullback (`vjp(bwd)`) that does not involve `property`; folding the iso into `property` would reference it twice and force a `State`-shaped cotangent seed. Not on the `Lens` protocol: `reparam` is a standalone optic the step differentiates, not a method on every lens.

## Example nits
- `to_real`/`to_fractional` are on `Frame`, not `Cell` → `cell.frame.to_real(...)`.
- The updated example is a `stop_gradient` straight-through: primal no-op (`new_pos == positions`), injecting only the *gradient* coupling: the cotangent job the scope section defers, not the primal normalization the body describes. The `Iso` framing unifies the two.

---

Related, split out of this discussion: #174 (composition must preserve custom setters; not a blocker here, the propagator uses `fwd`/`bwd` directly), #175 (lens `R` variance is unsound), #176 (optics naming/docstring cleanups).

**Comment by n-gao** [comment]
After internal discussion, we do want to go forward with this.
